plugins/USBPrinting/USBPrinterOutputDevice.py

```diff
@@ -198,26 +198,53 @@ def _updateFirmware(self):
 
         if not programmer.isConnected():
             Logger.log("e", "Unable to connect with serial. Could not update firmware")
-            self._updateFirmware_completed()
+            self._updateFirmware_completed_communication_error()
             return 
 
         self._updating_firmware = True
 
         try:
             programmer.programChip(hex_file)
             self._updating_firmware = False
+        except serial.SerialException as e:
+            Logger.log("e", "SerialException while trying to update firmware: <%s>" %(repr(e)))
+            self._updateFirmware_completed_io_error()
+            return
         except Exception as e:
-            Logger.log("e", "Exception while trying to update firmware %s" %e)
-            self._updateFirmware_completed()
+            Logger.log("e", "Exception while trying to update firmware: <%s>" %(repr(e)))
+            self._updateFirmware_completed_unknown()
             return
         programmer.close()
 
-        self._updateFirmware_completed()
+        self._updateFirmware_completed_sucessfully()
         return
 
-    ##  Private function which makes sure that firmware update process has completed/ended
-    def _updateFirmware_completed(self):
-        self.setProgress(100, 100)
+    ##  Private function which makes sure that firmware update process has failed by missing firmware
+    def _updateFirmware_completed_missing_firmware(self):
+        return self._updateFirmware_completed_common(progress = -4)
+    
+    ##  Private function which makes sure that firmware update process has failed by an IO error
+    def _updateFirmware_completed_io_error(self):
+        return self._updateFirmware_completed_common(progress = -3)
+    
+    ##  Private function which makes sure that firmware update process has failed by a communication problem
+    def _updateFirmware_completed_communication_error(self):
+        return self._updateFirmware_completed_common(progress = -2)
+    
+    ##  Private function which makes sure that firmware update process has failed by an unknown error
+    def _updateFirmware_completed_unknown(self):
+        return self._updateFirmware_completed_common(progress = -1)
+    
+    ##  Private function which makes sure that firmware update process has sucessfully completed/ended
+    def _updateFirmware_completed_sucessfully(self):
+        return self._updateFirmware_completed_common(progress = 100)
+    
+    ##  Private common function which makes sure that firmware update process has completed/ended with a set progress state
+    def _updateFirmware_completed_common(self, progress, max_progress = 100):
```

Comment:
Instead of masquerading an errorcode in progress it would be better to have an actual error argument here, and a @pyqtProperty named errorCode.